bot/utils/time.py

```diff
@@ -111,3 +120,64 @@ def parse_rfc1123(stamp: str) -> datetime.datetime:
 def format_infraction(timestamp: str) -> str:
     """Format an infraction timestamp to a more readable ISO 8601 format."""
     return dateutil.parser.isoparse(timestamp).strftime(INFRACTION_FORMAT)
+
+
+def get_duration(date_from: datetime.datetime, date_to: datetime.datetime, parts: Optional[int] = 2) -> str:
+    """
+    Get the duration between two datetime, in human readable format.
+
+    Will return <parts> number of units if avaiable, for example:
+    - 11 hours, 59 minutes
+    - 1 week, 6 minutes
+    - 7 months, 2 weeks
+    - 3 years, 3 months
+    - 5 minutes
+
+    :param date_from: A datetime.datetime object.
+    :param date_to: A datetime.datetime object.
+    :param parts: An int, defauted to two - the amount of units to return.
+    """
+    div = abs(date_from - date_to).total_seconds()
+    div = round(div, 0)  # to avoid (14 minutes, 60 seconds)
+    results: List[str] = []
+    for unit, name in TIME_MARKS:
+        div, amount = divmod(div, unit)
+        if amount > 0:
+            plural = 's' if amount > 1 else ''
+            results.append(f"{amount:.0f} {name}{plural}")
+    parts = parts if parts is not None else len(results)  # allow passing None directly to return all parts
+    # We have to reverse the order of units because currently it's smallest -> largest
+    return ', '.join(results[::-1][:parts])
+
+
+def get_duration_from_expiry(
+    expiry: str = None,
+    date_from: datetime.datetime = None,
+    parts: Optional[int] = 2
+) -> Optional[str]:
+    """
+    Get the duration between datetime.utcnow() and an expiry, in human readable format.
+
+    Will return the two biggest units avaiable, for example:
+    - 11 hours, 59 minutes
+    - 1 week, 6 minutes
+    - 7 months, 2 weeks
+    - 3 years, 3 months
+    - 5 minutes
+
+    :param expiry: A string. If not passed in, will early return a None ( Permanent infraction ).
+    :param date_from: A datetime.datetime object. If not passed in, will use datetime.utcnow().
+    :param parts: An int, to show how many parts will be returned ( year - month or year - month - week - day ...).
+    """
+    if not expiry:
+        return None
+
+    date_from = date_from or datetime.datetime.utcnow()
+    date_to = dateutil.parser.isoparse(expiry).replace(tzinfo=None)
+
+    expiry_formatted = format_infraction(expiry)
+
+    duration = get_duration(date_from, date_to, parts)
+    duration_formatted = f" ({duration})" if duration else ''
+
+    return f"{expiry_formatted}{duration_formatted}"
```

Comment:
I'm a bit confused by these functions since they seem to share a lot of functionality with the `humanize_delta` utility function that converts a `relativedelta` to a human-readable duration string. Why don't we just parse the datetime string, calculate the `dateutils.relativedelta` and use the existing function to get a human-readable duration?

Reply:
Oh my godness, you are correct, I did not realize we have a `humanize_delta` right there lol, and it works the same, if not more accurate across months and years!

I will probably just write an extra function to handle the expiry but call `humanize_delta` instead!